src/lib/net/SecureSocket.cpp

```diff
@@ -699,19 +667,30 @@ SecureSocket::disconnect()
     sendEvent(getEvents()->forIStream().inputShutdown());
 }
 
-bool SecureSocket::verify_cert_fingerprint(const barrier::fs::path& fingerprint_db_path)
+bool SecureSocket::verify_peer_certificate(const barrier::fs::path& fingerprint_db_path)
 {
     // ssl_mutex_ is assumed to be acquired
 
+    // ensure peer presented a certificate
+    X509* cert = SSL_get_peer_certificate(m_ssl->m_ssl);
+    if (cert == NULL) {
+        showError("peer has no ssl certificate");
+        return false;
+    }
+    char* line = X509_NAME_oneline(X509_get_subject_name(cert), 0, 0);
+    LOG((CLOG_INFO "peer ssl certificate info: %s", line));
+    OPENSSL_free(line);
+
     // calculate received certificate fingerprint
     barrier::FingerprintData fingerprint_sha1, fingerprint_sha256;
     try {
-        auto* cert = SSL_get_peer_certificate(m_ssl->m_ssl);
         fingerprint_sha1 = barrier::get_ssl_cert_fingerprint(cert,
-                                                             barrier::FingerprintType::SHA1);
+                                                                barrier::FingerprintType::SHA1);
         fingerprint_sha256 = barrier::get_ssl_cert_fingerprint(cert,
-                                                               barrier::FingerprintType::SHA256);
+                                                                barrier::FingerprintType::SHA256);
+        X509_free(cert);
```

Comment:
Instead of trying to add X509_free to all paths, it's better to automate this using RAII.

Add `auto cert_free = barrier::finally([cert]() { X509_free(cert); });` above immediately below when we know that cert needs to be freed.

Reply:
Sure. I've seen that barrier::finally, and understand its use. It just felt unnecessary in this case, with only two known, and very close branches.

Reply:
In the future code may evolve and especially in the presence of exceptions it's rather easy to introduce memory leaks.

Reply:
I agree that for the current code `finally` was probably not needed, I'm more focusing in how this could break in the future. In the long term it would be great if all such allocations could be freed by `finally`, then the chance of resource leaks would be greatly reduced.